cannon/mipsevm/versions/detect.go

```diff
@@ -0,0 +1,35 @@
+package versions
+
+import (
+	"fmt"
+	"io"
+
+	"github.com/ethereum-optimism/optimism/cannon/serialize"
+	"github.com/ethereum-optimism/optimism/op-service/ioutil"
+)
+
+func DetectVersion(path string) (StateVersion, error) {
+	if !serialize.IsBinaryFile(path) {
+		return VersionSingleThreaded, nil
+	}
+
+	var f io.ReadCloser
+	f, err := ioutil.OpenDecompressed(path)
+	if err != nil {
+		return 0, fmt.Errorf("failed to open file %q: %w", path, err)
+	}
+	defer f.Close()
+
+	var ver StateVersion
+	bin := serialize.NewBinaryReader(f)
+	if err := bin.ReadUInt(&ver); err != nil {
+		return 0, err
+	}
+
+	switch ver {
+	case VersionSingleThreaded, VersionMultiThreaded:
+		return ver, nil
+	default:
+		return 0, fmt.Errorf("%w: %d", ErrUnknownVersion, ver)
```

Comment:
Yeah. This is always a good practice. Unfortunately, doing this now will be breaking as we have state files that use version 0. But perhaps that's nbd if we can effectively regenerate those state files.

Reply:
I've decided to leave the single threaded version zero-valued. I think it's more trouble than it's worth to break state compat, especially since we'll be deprecating that version soon with Adrian's change.